### PR TITLE
APPSRE-3207: Generate `REPO_DIGEST`

### PR DESCRIPTION
**Pull request by 2uasimojo**
Support a new template parameter, `REPO_DIGEST`. If required in the template, we will query the registry data for the image based on `REGISTRY_IMG` and `IMAGE_TAG`, which must also be present (whether specified or generated). `REPO_DIGEST` will be set using the same form as produced by

`docker image inspect ${REGISTRY_IMG}:${IMAGE_TAG} --format '{{index .RepoDigests 0}}'`

(except the image need not be present locally)

That is, `${REGISTRY_IMG}@sha256${DIGEST}`, where `${DIGEST}` is a 64c
hex string. For example, given:
- `REGISTRY_IMG = "quay.io/app-sre/osd-metrics-exporter-registry"`
- `IMAGE_TAG` = "production-2cb14b0"`
...we will set:
- `REPO_DIGEST = "quay.io/app-sre/osd-metrics-exporter-registry@sha256:c94893c302cf5ef8aae0581f60bf59c3f65860aeefdd6046440bdc95f258c26e"`

APPSRE-3207

### CODE REVIEW

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [inline reconcile/utils/saasherder.py:391]: This delta not strictly necessary, just saved an indent and made the two chunks match stylistically.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [inline reconcile/utils/saasherder.py:406]: Where do I go to make sure the `docker` package is installed?

**Comment by jharrington22** [comment]
@2uasimojo looks good to me - shouldn't effect existing deploys unless OC templates include this parameter. 

To me IMAGE_DIGEST makes more sense than REPO_DIGEST since we're switching on uses digests for image references.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:391]: Nope. The second argument to `.get()` is the default value to use if the key is not present. The default default is `None`, but here I switched to use `{}` (the empty dict) which will make the loop iterate zero times.

File: reconcile/utils/saasherder.py
Comment (rrati) [reply reconcile/utils/saasherder.py:406]: https://github.com/app-sre/container-images

**Comment by 2uasimojo** [comment]
Okay, note to reviewers: The second commit is formatting only. Rather than manually editing to conform to flake8, I ran `black` on the thing. You should review only the first commit for content.

**Comment by jharrington22** [comment]
> > To me IMAGE_DIGEST makes more sense than REPO_DIGEST since we're switching on the use of digests for image references.
> 
> I'm matching the `docker inspect` field, which is (a list) called `RepoDigests`. (Only guessing, but that's probably short for "Repository URI plus image digest".)

Makes sense.

**Comment by 2uasimojo** [comment]
> Okay, note to reviewers: The second commit is formatting only. Rather than manually editing to conform to flake8, I ran `black` on the thing. You should review only the first commit for content.

Bleh, I forgot `black` uses line length as a suggestion. I'll go back and reformat by hand :(

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:406]: :+1: 

https://github.com/app-sre/container-images/pull/13

**Comment by 2uasimojo** [comment]
/hold

Depends-On: https://github.com/app-sre/container-images/pull/13

File: reconcile/utils/saasherder.py
Comment (apahim) [inline reconcile/utils/saasherder.py:407]: Are we sure that saas deployer has pull credentials in place to ANY images that can ever be templated?

Comment (apahim) [reply reconcile/utils/saasherder.py:407]: Maybe what we have to do is to keep IMAGE_TAG unchanged and introduce a new IMAGE_DIGEST tag that will trigger this special behavior.

Comment (apahim) [reply reconcile/utils/saasherder.py:407]: On top of that, we need to make sure we have at least the quay.io/app-sre pull credentials in the saas-deployer job env.

Comment (2uasimojo) [reply reconcile/utils/saasherder.py:407]: Probably not. Nor do I suppose we can count on that being the case across the board, since the push could technically happen anywhere, not necessarily in chain with this saas deployer run.

I'm going to say we should defer that problem to future, if/when we actually encounter it. In the meantime, if someone tries to request `REPO_DIGEST` for a private image, this will blow up, which is good. (I.e. we don't want it to silently succeed with some bogus value.)

File: reconcile/utils/saasherder.py
Comment (apahim) [inline reconcile/utils/saasherder.py:410]: We have a lib, already part of the qontract-reconcile, that can grab the digest:
```python
>>> from sretoolbox.container import Image
>>> 
>>> i = Image('quay.io/app-sre/qontract-reconcile:latest', username=None, password=None)
>>> i.url_digest
'quay.io/app-sre/qontract-reconcile@sha256:9d119edb4c906e013c1b72c97cdab3fd26b35bed96f9b436990f43b9dbf9780e'
```

Missing image:

```python
>>> i = Image('quay.io/app-sre/missing:latest', username=None, password=None)
>>> i.url_digest
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/apahim/git/sretoolbox/sretoolbox/container/image.py", line 336, in url_digest
    url_digest += f'/{self.image}@{self.digest}'
  File "/home/apahim/git/sretoolbox/sretoolbox/container/image.py", line 79, in digest
    manifest = self._get_manifest()
  File "/home/apahim/git/sretoolbox/sretoolbox/container/image.py", line 94, in _get_manifest
    return self._request_get(url)
  File "/home/apahim/git/sretoolbox/sretoolbox/utils/retry.py", line 28, in f_retry
    defaults to ()
  File "/home/apahim/git/sretoolbox/sretoolbox/utils/retry.py", line 23, in f_retry
    :type exceptions: tuple(Exception, ...), optional
  File "/home/apahim/git/sretoolbox/sretoolbox/container/image.py", line 312, in _request_get
    self._raise_for_status(response)
  File "/home/apahim/git/sretoolbox/sretoolbox/container/image.py", line 286, in _raise_for_status
    raise HTTPError(msg)
requests.exceptions.HTTPError: (401) UNAUTHORIZED, access to the requested resource is not authorized
```

File: reconcile/utils/saasherder.py
Comment (apahim) [inline reconcile/utils/saasherder.py:416]: Let's be specific about the exceptions we are handling here.

File: reconcile/utils/saasherder.py
Comment (apahim) [reply reconcile/utils/saasherder.py:410]: Not asking you to use it, just presenting you with an option that's already part of the repo :)

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:407]: > Maybe what we have to do is to keep IMAGE_TAG unchanged and introduce a new IMAGE_DIGEST tag that will trigger this special behavior.

That's exactly what I've done (except `s/IMAGE_DIGEST/REPO_DIGEST/`).

> On top of that, we need to make sure we have at least the quay.io/app-sre pull credentials in the saas-deployer job env.

I mean, we _can_, but perhaps that's future work? I can make a note (where?) that this currently only works for public images.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:416]: I can peel out `KeyError`, but I'd like to keep the catch-all in any case, because I don't know what kinds of exceptions the docker lib can raise. Acceptable? (There are other spots in this file that use `Exception`...)

Comment (apahim) [reply reconcile/utils/saasherder.py:416]: We are cleaning up the pylint errors, generic exceptions being one of them. Having this here now only increases the upcoming amount of work.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:410]: Neat! And it doesn't require the `docker` package. I'll switch it.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:416]: Understood. If I miss one, it'll blow up ugly instead of pretty -- I assume that's an acceptable risk?

File: reconcile/utils/saasherder.py
Comment (jharrington22) [reply reconcile/utils/saasherder.py:410]: I'm assuming is assuming the registry is up. Whats the exception if the registry is down?

**Comment by jharrington22** [comment]
@rrati @maorfr I see this has a hold, I think we need to make an internal decision before this gets merged.

File: reconcile/utils/saasherder.py
Comment (apahim) [reply reconcile/utils/saasherder.py:410]: `requests.exceptions.ConnectionError`

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:407]: Update here: I've switched to using `sretoolbox.container.Image` and funneled in the same `image_auth` used elsewhere. So this should be good as far as creds already exist. And I believe that also puts the onus on the client to make sure they have the right creds in place if necessary, right?

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py]: This is done.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py]: This is done.

File: reconcile/utils/saasherder.py
Comment (maorfr) [inline reconcile/utils/saasherder.py:407]: we generally don't log info, as we look at the output of the integration and only want to see actions that occured. i suggest to change this to debug

File: reconcile/utils/saasherder.py
Comment (maorfr) [inline reconcile/utils/saasherder.py:409]: sorry to be petty, but can you name variables in the same way we do? `image_tag` over `img_tag`, and no abbreviations in general

File: reconcile/utils/saasherder.py
Comment (maorfr) [inline reconcile/utils/saasherder.py:414]: IMAGE_TAG will never be missing

File: reconcile/utils/saasherder.py
Comment (maorfr) [inline reconcile/utils/saasherder.py:403]: can we use "IMAGE_DIGEST" instead?

File: reconcile/utils/saasherder.py
Comment (maorfr) [inline reconcile/utils/saasherder.py:639]: :heart:

**Review by maorfr — CHANGES_REQUESTED** [review]
thanks a lot for this contribution!

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:403]: I chose `REPO_DIGEST` to match how `docker image inspect` names the corresponding field (`RepoDigests`, a list) in their JSON output. (I'm assuming that's short for "Repository URI plus image digest".)

Comment (maorfr) [reply reconcile/utils/saasherder.py:403]: good thing we are not using docker anymore?

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py]: Reverted.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py]: Done.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py]: Done.

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:417]: Based on L392-394 -- and reinforced by the comment on L395 -- it looks like it can be missing if it's not included in the saas parameters or the template parameters. Is there somewhere else that it's magically injected in that case?

File: reconcile/utils/saasherder.py
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:405]: Okay, but I agree with them on this point: The "image digest" is the `sha256:{hash}` part, not the whole URI. So IMO `IMAGE_DIGEST` isn't the correct name. And `IMAGE_URI_BY_DIGEST` gets a bit mealy, especially with python line width restrictions enforced at this level of indent :)

File: reconcile/utils/saasherder.py
Comment (maorfr) [reply reconcile/utils/saasherder.py:417]: Ah you are right

File: reconcile/utils/saasherder.py
Comment (maorfr) [reply reconcile/utils/saasherder.py:405]: Fair enough

**Review by maorfr — APPROVED** [review]
LGTM!

Thanks again. The result is very clear and to the point. I'll merge and promote with some docs tomorrow.

**Pull request merged**
